### PR TITLE
[BEAM-1551] Allow `PAssert`s to take a message

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by aviemzur** [comment]
R: @tgroh

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java [hunk 1]
```diff
@@ -332,10 +368,23 @@ private PAssert() {}
    * {@code Coder<K, V>}.
    */
   public static <K, V> SingletonAssert<Map<K, V>> thatMap(PCollection<KV<K, V>> actual) {
+    return thatMap(actual.getName(), actual);
+  }
+
+  /**
+   * Constructs a {@link SingletonAssert} for the value of the provided {@link PCollection}, which
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java]: `for the value of the provided {@link PCollection} with the specified reason. The {@link PCollection} must have at most one value per key.`

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java [hunk 2]
```diff
@@ -385,10 +385,11 @@ public void testContainsInAnyOrderFalse() throws Exception {
   @Category(RunnableOnService.class)
   public void testEmptyFalse() throws Exception {
     PCollection<Long> vals = pipeline.apply(CountingInput.upTo(5L));
-    PAssert.that(vals).empty();
+    PAssert.that("Vals should have been empty", vals).empty();
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java]: Add the same to a view-based assertion; I believe we already have tests for them.

It's also would be nice to ensure that by default the name of the input PCollection is present.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java [hunk 3]
```diff
@@ -316,12 +340,24 @@ private PAssert() {}
    */
   public static <K, V> SingletonAssert<Map<K, Iterable<V>>> thatMultimap(
       PCollection<KV<K, V>> actual) {
+    return thatMultimap(actual.getName(), actual);
+  }
+
+  /**
+   * Constructs a {@link SingletonAssert} for the value of the provided {@link PCollection}.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java]: "... with the specified reason."

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java [hunk 4]
```diff
@@ -297,15 +313,23 @@ private PAssert() {}
...
    * {@code PCollection PCollection<T>}, which must be a singleton.
    */
   public static <T> SingletonAssert<T> thatSingleton(PCollection<T> actual) {
-    return new PCollectionViewAssert<>(actual, View.<T>asSingleton(), actual.getCoder());
+    return thatSingleton(actual.getName(), actual);
+  }
+
+  /**
+   * Constructs a {@link SingletonAssert} for the value of the provided
+   * {@code PCollection PCollection<T>}, which must be a singleton.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java]: PCollection} with the specified reason. The provided PCollection must be a singleton.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java [hunk 5]
```diff
@@ -285,6 +292,15 @@ private PAssert() {}
    */
   public static <T> IterableAssert<T> thatSingletonIterable(
       PCollection<? extends Iterable<T>> actual) {
+    return thatSingletonIterable(actual.getName(), actual);
+  }
+
+  /**
+   * Constructs an {@link IterableAssert} for the value of the provided {@link PCollection} which
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java]: with the specified reason. The provided PCollection must contain...

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java [hunk 6]
```diff
@@ -276,7 +276,14 @@ private PAssert() {}
    * Constructs an {@link IterableAssert} for the elements of the provided {@link PCollection}.
    */
   public static <T> IterableAssert<T> that(PCollection<T> actual) {
-    return new PCollectionContentsAssert<>(actual);
+    return that(actual.getName(), actual);
+  }
+
+  /**
+   * Constructs an {@link IterableAssert} for the elements of the provided {@link PCollection}.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java]: PCollection} with the specified reason.

**Comment by tgroh** [comment]
Run Flink RunnableOnService

**Comment by aviemzur** [comment]
@tgroh addressed your comments in my latest 3 commits.

**Comment by tgroh** [comment]
Run Flink RunnableOnService

**Comment by tgroh** [comment]
Run Dataflow RunnableOnService

**Comment by tgroh** [comment]
LGTM, but the flink failures do look related. Could you take a look?

**Comment by aviemzur** [comment]
Run Apex RunnableOnService

**Comment by aviemzur** [comment]
Run Spark RunnableOnService

**Comment by aviemzur** [comment]
Run Gearpump RunnableOnService

**Comment by aviemzur** [comment]
Run Dataflow RunnableOnService

**Comment by aviemzur** [comment]
Run Flink RunnableOnService

**Comment by aviemzur** [comment]
Run Dataflow RunnableOnService

**Comment by aviemzur** [comment]
Run Spark RunnableOnService

**Comment by aviemzur** [comment]
Run Apex RunnableOnService

**Comment by aviemzur** [comment]
Run Spark RunnableOnService

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java [hunk 7]
```diff
@@ -1035,15 +1101,8 @@ private SideInputCheckerDoFn(
 
     @ProcessElement
     public void processElement(ProcessContext c) {
-      try {
```
Comment (aviemzur) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java]: This suppression of exceptions was the reason the side input tests failed in Flink runner running in streaming mode.
Removing this solves the issue. All the ROS tests pass now (Including the new ones).
@kennknowles PTAL

**Comment by aviemzur** [comment]
@tgroh committed a fix to solve the failing tests. It was an issue with `PAssert` itself.

**Comment by kennknowles** [comment]
LGTM. I'll merge it.

**Pull request merged**